File: 전예진/DP/prob1.py
def solution(N, number):
    if N == number: # N과 number가 같으면 1회로 표현 가능
        return 1
    
    # DP를 위한 집합 초기화
    dp = [set() for _ in range(9)]
    
    # 각 집합에 각각의 N을 연속해서 나열한 경우 초기화
    for i in range(1, 9):
        dp[i].add(int(str(N) * i))
    
    # DP 진행
    for i in range(1, 9):
        for j in range(1, i):
            for num1 in dp[j]:
                for num2 in dp[i - j]:
                    dp[i].add(num1 + num2)
                    dp[i].add(num1 - num2)
                    dp[i].add(num1 * num2)
                    if num2 != 0:
                        dp[i].add(num1 // num2)
        
        if number in dp[i]:
            return i
    
    return -1 

File: 주다애/DP/prob1.java
import java.util.*;

class Solution {
    // 중복 저장할 필요 없음
    static List<Set<Integer>> set = new ArrayList<>();
    public int solution(int N, int number) {
        int answer = 0;
        for (int i = 0; i <= 8; i++) {
            set.add(new HashSet<>());             
        }
        set.get(1).add(N);
        
        // 계산할 필요 없음
        if(N == number) {
            return 1;
        }
        
        String st = Integer.toString(N);
        for(int i = 2; i <= 8; i++) {
            for(int j = 1; j <= i / 2; j++) {
                operations(i, set.get(i - j), set.get(j));
                operations(i, set.get(j), set.get(i - j));
            }
            // 55, 555와 같은 연속되는 숫자 저장
            st += Integer.toString(N);
            set.get(i).add(Integer.valueOf(st));  
            
            if(set.get(i).contains(number)) {
                return i;
            }
        }
        return -1;
    }
    
    static void operations(int i, Set<Integer> first, Set<Integer> second) {
        for(int a : first) {
            for(int b : second) {
                set.get(i).add(a + b);
                set.get(i).add(a - b);
                set.get(i).add(a * b);
                if(b != 0) {
                    set.get(i).add(a / b);
                }
            }
        }
    }
}
